packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -682,160 +948,607 @@ class _SortData {
+///       ),
+///     );
+///   }
+/// }
+/// ```
+///
+/// ```dart
+/// Widget build(BuildContext context) {
+///   return FocusTraversalGroup(
+///     policy: OrderedTraversalPolicy(),
+///     child: Column(
+///       mainAxisAlignment: MainAxisAlignment.center,
+///       children: <Widget>[
+///         Row(
+///           mainAxisAlignment: MainAxisAlignment.center,
+///           children: const <Widget>[
+///             DemoButton(name: 'Six', order: 6),
+///           ],
+///         ),
+///         Row(
+///           mainAxisAlignment: MainAxisAlignment.center,
+///           children: const <Widget>[
+///             DemoButton(name: 'Five', order: 5),
+///             DemoButton(name: 'Four', order: 4),
+///           ],
+///         ),
+///         Row(
+///           mainAxisAlignment: MainAxisAlignment.center,
+///           children: const <Widget>[
+///             DemoButton(name: 'Three', order: 3),
+///             DemoButton(name: 'Two', order: 2),
+///             DemoButton(name: 'One', order: 1, autofocus: true),
+///           ],
+///         ),
+///       ],
+///     ),
+///   );
+/// }
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [WidgetOrderFocusTraversalPolicy], a policy that relies on the widget
+///    creation order to describe the order of traversal.
+///  * [ReadingOrderTraversalPolicy], a policy that describes the order as the
+///    natural "reading order" for the current [Directionality].
+///  * [NumericFocusOrder], a focus order that assigns a numeric traversal order
+///    to a [FocusTraversalOrder] widget.
+///  * [LexicalFocusOrder], a focus order that assigns a string-based lexical
+///    traversal order to a [FocusTraversalOrder] widget.
+///  * [FocusOrder], an abstract base class for all types of focus traversal
+///    orderings.
+class OrderedTraversalPolicy extends FocusTraversalPolicy with DirectionalFocusTraversalPolicyMixin {
+  /// Constructs a traversal policy that orders widgets for keyboard traversal
+  /// based on an explicit order.
+  ///
+  /// If [secondary] is null, it will default to [ReadingOrderTraversalPolicy].
+  OrderedTraversalPolicy({this.secondary});
+
+  /// This is the policy that is used when a node doesn't have an order
+  /// assigned, or when multiple nodes have orders which are identical.
+  ///
+  /// If not set, this defaults to [ReadingOrderTraversalPolicy].
+  ///
+  /// This policy determines the secondary sorting order of nodes which evaluate
+  /// as having an identical order (including those with no order specified).
+  ///
+  /// Nodes with no order specified will be sorted after nodes with an explicit
+  /// order.
+  final FocusTraversalPolicy secondary;
+
+  @override
+  Iterable<FocusNode> sortDescendants(Iterable<FocusNode> descendants) {
+    final FocusTraversalPolicy secondaryPolicy = secondary ?? ReadingOrderTraversalPolicy();
+    final Iterable<FocusNode> sortedDescendants = secondaryPolicy.sortDescendants(descendants);
```

Comment:
I wish we wouldn't have to sort according to the secondary policy if there's no need to do so. I imagine in most use cases, you would fully specify a numeric sort order for most (maybe even all) nodes?

Reply:
Well, but if even two nodes compare as "equal", we have to give them some kind of deterministic order.  Also, imagine mixing your own code on a page with a widget that someone else provides: even if you provide a sort order for your widgets, the other one won't have done it the same way (and probably didn't provide an order for each one).  You can order their group of widgets within yours with a group, but if you don't do that (or don't know to do that), it should be reasonable and deterministic.